File: src/java/engine/movegen/bitboards/BishopMoveGenerator.java
package java.engine.movegen.bitboards;

import java.util.function.BiFunction;
import java.util.stream.Stream;

/**
 * Created by melvic on 7/29/18.
 */
public interface BishopMoveGenerator extends SlidingMoveGenerator {
    public default Stream<BiFunction<Integer, Long, Long>> getAllMoves() {
        return Stream.of(this::positiveDiagonal,
                this::negativeDiagonal,
                this::positiveAntiDiagonal,
                this::negativeAntiDiagonal);
    }

    public default long positiveDiagonal(int bishopPosition, long occupied) {
        return ray(bishopPosition, occupied, this::positiveSlide, this::getDiagonalMask);
    }

    public default long negativeDiagonal(int bishopPosition, long occupied) {
        return ray(bishopPosition, occupied, this::negativeSlide, this::getDiagonalMask);
    }

    public default long positiveAntiDiagonal(int bishopPosition, long occupied) {
        return ray(bishopPosition, occupied, this::positiveSlide, this::getAntiDiagonalMask);
    }

    public default long negativeAntiDiagonal(int bishopPosition, long occupied) {
        return ray(bishopPosition, occupied, this::negativeSlide, this::getAntiDiagonalMask);
    }
}

File: src/engine/movegen/bitboards/BishopMoveGenerator.scala
package engine.movegen.bitboards

/**
  * Created by melvic on 9/23/18.
  */
object BishopMoveGenerator extends SlidingMoveGenerator {
  def positiveDiagonal: Slide = positiveRay(diagonalMask)

  def negativeDiagonal: Slide = negativeRay(diagonalMask)

  def positiveAntiDiagonal: Slide = positiveRay(antiDiagonalMask)

  def negativeAntiDiagonal: Slide = negativeRay(antiDiagonalMask)

  override def moves = Stream(positiveDiagonal,
    negativeDiagonal, positiveAntiDiagonal, negativeAntiDiagonal)
}
